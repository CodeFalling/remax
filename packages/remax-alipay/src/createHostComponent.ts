import * as React from 'react';

export default function createHostComponent<P = any>(name: string) {
  const Component: React.ForwardRefRenderFunction<any, P> = (
    props,
    ref: React.Ref<any>
  ) => {
    const { children = [] } = props;
    return React.createElement(name, { ...props, ref }, children);
  };
  return React.forwardRef<any, React.PropsWithChildren<P>>(Component);
}
