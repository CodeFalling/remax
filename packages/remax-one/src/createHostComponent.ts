import * as React from 'react';
import {
  TapEvent,
  TouchEvent,
  EventTarget,
  EventCurrentTarget,
  ImageLoadEvent,
  ImageErrorEvent,
  InputEvent,
  FormEvent,
} from './types';

export function createTarget(target: any, detail: any): EventTarget {
  return {
    id: target.id,
    offsetLeft: target.offsetLeft,
    offsetTop: target.offsetTop,
    dataset: target.targetDataset || target.dataset,
    value: detail?.value,
  };
}

export function createCurrentTarget(currentTarget: any): EventCurrentTarget {
  return {
    offsetLeft: currentTarget.offsetLeft,
    offsetTop: currentTarget.offsetTop,
    dataset: currentTarget.dataset,
  };
}

export const createTapEvent = (originalEvent: any): TapEvent => ({
  type: originalEvent.type,
  stopPropagation: originalEvent.stopPropagation,
  target: createTarget(originalEvent.target, originalEvent.detail),
  currentTarget: createCurrentTarget(originalEvent.currentTarget),
  originalEvent,
});

export const createTouchEvent = (originalEvent: any): TouchEvent => ({
  type: originalEvent.type,
  target: createTarget(originalEvent.target, originalEvent.detail),
  currentTarget: createCurrentTarget(originalEvent.currentTarget),
  touches: originalEvent.touches,
  changedTouches: originalEvent.touches,
  originalEvent,
});

export const createImageEvent = (originalEvent: any): ImageLoadEvent | ImageErrorEvent => ({
  type: originalEvent.type,
  target: createTarget(originalEvent.target, originalEvent.detail),
  currentTarget: createCurrentTarget(originalEvent.currentTarget),
  originalEvent,
});

export function createCallback(fn: Function | undefined, eventCreator: Function) {
  if (typeof fn !== 'function') {
    return undefined;
  }

  return (originalEvent: any) => fn(eventCreator(originalEvent));
}

export const createInputEvent = (originalEvent: any): InputEvent => ({
  type: originalEvent.type,
  target: createTarget(originalEvent.target, originalEvent.detail),
  currentTarget: createCurrentTarget(originalEvent.currentTarget),
  originalEvent,
});

export const createFormEvent = (originalEvent: any): FormEvent => ({
  type: originalEvent.type,
  target: createTarget(originalEvent.target, originalEvent.detail),
  currentTarget: createCurrentTarget(originalEvent.currentTarget),
  originalEvent,
});

export default function createHostComponent<P = any>(name: string) {
  const Component: React.ForwardRefRenderFunction<any, P> = (props: any, ref: React.Ref<any>) => {
    const inputProps = {
      ...props,
      onLongTap: createCallback(props.onLongTap, createTapEvent),
      onTap: createCallback(props.onTap, createTapEvent),
      onTouchStart: createCallback(props.onTouchStart, createTouchEvent),
      onTouchMove: createCallback(props.onTouchMove, createTouchEvent),
      onTouchEnd: createCallback(props.onTouchEnd, createTouchEvent),
      onTouchCancel: createCallback(props.onTouchCancel, createTouchEvent),
      onChange: createCallback(props.onChange, createInputEvent),
      onInput: createCallback(props.onInput, createInputEvent),
      onConfirm: createCallback(props.onConfirm, createInputEvent),
      onFocus: createCallback(props.onFocus, createInputEvent),
      onBlur: createCallback(props.onBlur, createInputEvent),
      onSubmit: createCallback(props.onSubmit, createFormEvent),
      onReset: createCallback(props.onReset, createFormEvent),
    };

    if (name === 'image') {
      inputProps.onLoad = createCallback(props.onLoad, createImageEvent);
      inputProps.onError = createCallback(props.onError, createImageEvent);
    }

    return React.createElement(name, { ...inputProps, ref });
  };
  return React.forwardRef<any, React.PropsWithChildren<P>>(Component);
}
