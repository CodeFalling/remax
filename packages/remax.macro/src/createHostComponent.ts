import * as t from '@babel/types';
import { NodePath } from '@babel/traverse';

const PACKAGE_NAME = 'remax';
const FUNCTION_NAME = 'createHostComponent';

export const hostComponents = new Map<
  string,
  { props: string[]; additional?: boolean }
>();

function getConfig(callExpression: NodePath<t.CallExpression>) {
  const args = callExpression.node.arguments;
  const name = args[0] as t.StringLiteral;
  const props = (args[1] as t.ArrayExpression).elements
    .filter(element => element !== null)
    .map(element => (element as t.StringLiteral).value);

  return [name, props];
}

function insertImportDeclaration(program: any) {
  let name = FUNCTION_NAME;

  const node = program.node.body.find(
    (node: t.Node) =>
      t.isImportDeclaration(node) && node.source.value === PACKAGE_NAME
  );

  if (node) {
    const specifier = (node as t.ImportDeclaration).specifiers.find(
      specifier =>
        t.isImportSpecifier(specifier) &&
        specifier.imported.name === FUNCTION_NAME
    );

    if (specifier) {
      name = specifier.local.name;
    } else {
      node.specifiers.push(
        t.importSpecifier(t.identifier(name), t.identifier(name))
      );
    }
  } else {
    const importDeclaration = t.importDeclaration(
      [t.importSpecifier(t.identifier(name), t.identifier(name))],
      t.stringLiteral(PACKAGE_NAME)
    );
    program.node.body.unshift(importDeclaration);
  }

  return name;
}

export default function createHostComponent(path: NodePath, state: any) {
  const program = state.file.path;
  const functionName = insertImportDeclaration(program);
  const callExpression = path.findParent(p =>
    t.isCallExpression(p)
  ) as NodePath<t.CallExpression>;
  const config = getConfig(callExpression);
  const name = config[0] as t.StringLiteral;
  const props = config[1] as string[];

  hostComponents.set(name.value, {
    props,
    additional: true,
  });

  callExpression.replaceWith(
    t.callExpression(t.identifier(functionName), [name])
  );
}
