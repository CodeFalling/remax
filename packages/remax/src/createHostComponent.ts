import * as React from 'react';
import { TYPE_TEXT } from './constants';

export const hostComponents = new Set();

export function isHostComponent(type: string) {
  if (type === TYPE_TEXT) {
    return true;
  }

  return hostComponents.has(type);
}

export default function createHostComponent<P = any>(
  name: string,
  component?: React.ComponentType<P>
) {
  hostComponents.add(name);

  if (component) {
    return component;
  }

  const Component: React.FC<P> = (props, ref: any) => {
    const { children = [] } = props;
    return React.createElement(name, { ...props, ref }, children);
  };
  return React.forwardRef<{}, React.PropsWithChildren<P>>(Component);
}
