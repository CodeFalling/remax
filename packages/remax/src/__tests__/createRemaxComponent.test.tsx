import * as React from 'react';
import TestRenderer from 'react-test-renderer';
import createHostComponent from '../createHostComponent';

describe('createHostComponent', () => {
  it('create remax component', () => {
    const RemaxCustomComponent = createHostComponent<{ foo: string }>(
      'custom-component'
    );
    const testRenderer = TestRenderer.create(
      <RemaxCustomComponent foo="bar" />
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
