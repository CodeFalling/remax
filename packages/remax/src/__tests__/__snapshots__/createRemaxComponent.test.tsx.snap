// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createHostComponent create remax component 1`] = `
<custom-component
  foo="bar"
/>
`;
